mesonbuild/environment.py

```diff
@@ -834,12 +834,23 @@ def detect_d_compiler(self, want_cross):
             raise EnvironmentException('Could not execute D compiler "%s"' % ' '.join(exelist))
         version = search_version(out)
         full_version = out.split('\n', 1)[0]
+
+        is_win64 = False
+        is_msvc = False
+        if mesonlib.is_windows():
+            is_msvc = 'VCINSTALLDIR' in os.environ
+            # detect which MSVC build environment is currently active
+            if is_msvc and detect_cpu_family({'c': self.detect_c_compiler(want_cross)}) == 'x86_64':
+                is_win64 = True
+            elif not is_msvc and detect_cpu_family({}) == 'x86_64':
+                is_win64 = True
+
         if 'LLVM D compiler' in out:
-            return compilers.LLVMDCompiler(exelist, version, is_cross, full_version=full_version)
+            return compilers.LLVMDCompiler(exelist, version, is_cross, is_win64, full_version=full_version)
         elif 'gdc' in out:
-            return compilers.GnuDCompiler(exelist, version, is_cross, full_version=full_version)
+            return compilers.GnuDCompiler(exelist, version, is_cross, is_win64, full_version=full_version)
         elif 'The D Language Foundation' in out or 'Digital Mars' in out:
-            return compilers.DmdDCompiler(exelist, version, is_cross, full_version=full_version)
+            return compilers.DmdDCompiler(exelist, version, is_cross, is_win64, is_msvc, full_version=full_version)
```

Comment:
You can't use the LLVM D compiler with MSVC?

Reply:
This line refers to DMD, not LLVM D compiler. All DMD builds (32-bit or 64-bit) on Windows always compiles to 32-bit COFF by default regardless of the environment variables, so we need to know if we are in 64-bit OS, and in MSVC environment for the compiler to target at the right architecture.

As for LLVM D compiler, the compiler does use native MSVC linker if the environment is set up, but it does not use the 32-bit linker when in 32-bit MSVC environment, so this requires adjusting as well.